scripts/test-pipelines.sh

```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env bash
+
+shopt -s globstar
+shopt -s nullglob
+
+# Quote the heredoc to prevent shell expansion
+cat << "EOF"
+workflow:
+  rules:
+    # Disable merge request pipelines
+    - if: $CI_MERGE_REQUEST_ID
+      when: never
+    - when: always
+default:
+  interruptible: true
+variables:
+  GH_PROJECT_PATH: "MatrixAI/${CI_PROJECT_NAME}"
+  GH_PROJECT_URL: "https://${GITHUB_TOKEN}@github.com/${GH_PROJECT_PATH}.git"
+  GIT_SUBMODULE_STRATEGY: "recursive"
+  # Cache .npm
+  NPM_CONFIG_CACHE: "./tmp/npm"
+  # Prefer offline node module installation
+  NPM_CONFIG_PREFER_OFFLINE: "true"
+  # `ts-node` has its own cache
+  # It must use an absolute path, otherwise ts-node calls will CWD
+  TS_CACHED_TRANSPILE_CACHE: "${CI_PROJECT_DIR}/tmp/ts-node-cache"
+  TS_CACHED_TRANSPILE_PORTABLE: "true"
+# Cached directories shared between jobs & pipelines per-branch per-runner
+cache:
+  key: $CI_COMMIT_REF_SLUG
+  paths:
+    - ./tmp/npm/
+    - ./tmp/ts-node-cache/
+    # `jest` cache is configured in jest.config.js
+    - ./tmp/jest/
+stages:
+  - check       # Linting, unit tests
+image: registry.gitlab.com/matrixai/engineering/maintenance/gitlab-runner
+EOF
+
+printf "\n"
+
+# Using shards to optimise tests
+# In the future we can incorporate test durations rather than using
+# a static value for the parallel keyword
+cat << "EOF"
+check:test:
+  stage: check
+  needs: []
+  parallel: 2
```

Comment:
This `parallel: 2` should be put up top as a big variable. You want it to be configurable easily.

Set it has `CI_PARALLEL_JOBS`. And give it a comment indicating how it translates to the shard count for the jest tests.

Reply:
Variables aren't converted to values when the yaml file is created it seems, so gitlab says the created file is invalid since the value for `parallel` needs to be an integer or a hash

```yml
check:test:
  stage: check
  needs: []
  parallel: $CI_PARALLEL
```

Reply:
There are 2 ways to do this, using gitlab's variables, which may not work for whatever reason, or embedding variables from the shell script, which should work, you just need a way to antiquote/quote any variables that should not be expanded. Right now `<<"EOF"` prevents shell expansion, so `$...` is interpreted literally. We may be able to break that apart, and allow interpolation for just that section of `parallel: $CI_PARALLEL`.

So I think `$CI_PARALLEL` should be set as a variable in the shell script, not the yaml file.

See: https://gist.github.com/CMCDragonkai/451d2c2b3d36a0eaf350

Reply:
Fixed by just separating that line out in the shell scripts like so
```sh
cat << "EOF"
check:test:
  stage: check
  needs: []
EOF
cat << EOF
  parallel: $CI_PARALLEL
EOF
cat << "EOF"
  script:
  ...
```